src/Features/Core/Portable/CodeRefactorings/MoveType/AbstractMoveTypeService.cs

```diff
@@ -199,58 +173,87 @@ private static bool AnyTopLevelTypeMatchesDocumentName(State state, Cancellation
+                return false;
         }
 
-        return namesMatch;
+        return true;
+    }
+
+    private static bool TypeNameMatches(string documentNameWithoutExtension, string typeName, int arity)
+    {
+        if (typeName.Equals(documentNameWithoutExtension, StringComparison.CurrentCulture))
+            return true;
+
+        if ($"{typeName}`{arity}".Equals(documentNameWithoutExtension, StringComparison.CurrentCulture))
+            return true;
+
+        return false;
     }
 
-    private static ImmutableArray<string> GetSuggestedFileNames(
+    private ImmutableArray<string> GetSuggestedFileNames(
         TTypeDeclarationSyntax typeNode,
-        bool isNestedType,
-        string typeName,
         string documentNameWithExtension,
-        SemanticModel semanticModel,
-        CancellationToken cancellationToken)
+        bool includeArity)
     {
+        var isNestedType = IsNestedType(typeNode);
+        var (typeName, arity) = this.GetSymbolNameAndArity(typeNode);
         var fileExtension = Path.GetExtension(documentNameWithExtension);
 
         var standaloneName = typeName + fileExtension;
 
-        // If it is a nested type, we should match type hierarchy's name parts with the file name.
+        using var _ = ArrayBuilder<string>.GetInstance(out var suggestedFileNames);
+
+        suggestedFileNames.Add(typeName + fileExtension);
+        if (includeArity && arity > 0)
+            suggestedFileNames.Add($"{typeName}`{arity}{fileExtension}");
+
         if (isNestedType)
         {
-            var typeNameParts = GetTypeNamePartsForNestedTypeNode(typeNode, semanticModel, cancellationToken);
-            var dottedName = typeNameParts.Join(".") + fileExtension;
+            var typeNameParts = GetTypeNamePartsForNestedTypeNode(typeNode);
+            AddNameParts(typeNameParts.Select(t => t.name));
 
-            return [standaloneName, dottedName];
+            if (includeArity && typeNameParts.Any(t => t.arity > 0))
+                AddNameParts(typeNameParts.Select(t => t.arity > 0 ? $"{t.name}`{t.arity}" : t.name));
         }
-        else
+
+        return suggestedFileNames.ToImmutableAndClear();
+
+        void AddNameParts(IEnumerable<string> parts)
+        {
+            AddNamePartsWithSeparator(parts, ".");
+            AddNamePartsWithSeparator(parts, "+");
+        }
+
+        void AddNamePartsWithSeparator(IEnumerable<string> parts, string separator)
         {
-            return [standaloneName];
+            suggestedFileNames.Add(parts.Join(separator) + fileExtension);
         }
     }
 
-    private static IEnumerable<string> GetTypeNamePartsForNestedTypeNode(
-        TTypeDeclarationSyntax typeNode, SemanticModel semanticModel, CancellationToken cancellationToken)
-            => typeNode.AncestorsAndSelf()
-                    .OfType<TTypeDeclarationSyntax>()
-                    .Select(n => semanticModel.GetRequiredDeclaredSymbol(n, cancellationToken).Name)
```

Comment:
no need to go to semantics here.  we have type-decl nodes, se we can always trivially figure out the type's name purely syntactically.